Add class for custom transforms to adapter, similar to `LambdaTransform`.

vpratz opened the pull request:
Closes #362.

This PR reintroduces the features that were present in `LambdaTransform`, but only allowing registered functions. While being stricter, this allows for closer scaffolding and raising errors early on, so that users cannot provide functions that will not be (de)serializable later on.

As there are a few failure modes, the focus is on providing detailed error messages to enable users to solve problems without external help.

Open questions:

- When writing the code, I chose the naming to make it very clear that you cannot pass any functions, but only serializable ones. In my impression, I overdid it a bit, and the point should be sufficiently clear, even when we rename `serializable_forward/inverse_fn` to `forward/inverse` again. What do you think?
- I opted for a method in the `Adapter`, as it helps in easily providing the user with the `FilterTransform` functionality. Is this ok with you, @LarsKue 
- As always, naming suggestions are welcome.

LarsKue commented on tests/test_adapters/test_adapters.py:
```diff
@@ -110,3 +111,71 @@ def test_simple_transforms(random_data):
+    # method instead of function provided
+    with pytest.raises(ValueError):
+        SerializableCustomTransform(serializable_forward_fn=A.fn, serializable_inverse_fn=registered_fn)
+        SerializableCustomTransform(serializable_forward_fn=registered_fn, serializable_inverse_fn=A.fn)
```
These need to be in separate `with` blocks. The block exits on the first `ValueError` that is raised.

vpratz replied:
Good spot, thanks!

LarsKue commented on bayesflow/adapters/adapter.py:
```diff
@@ -283,6 +284,88 @@ def apply(
+        include: str | Sequence[str] = None,
+        *,
+        serializable_forward_fn: Callable[[np.ndarray, ...], np.ndarray],
+        serializable_inverse_fn: Callable[[np.ndarray, ...], np.ndarray],
```
I agree that these are a bit verbose. The intent is clear from the name of the function and the docs, I think. Let's go with your suggestion to rename them back to `forward` and `inverse`.

LarsKue commented on bayesflow/adapters/transforms/serializable_custom_transform.py:
````diff
@@ -0,0 +1,184 @@
+def my_{label}(...):
+    [your code goes here...]
+```
+"""
````
Triple-quoted strings lead to some super weird indenting here. Can we move to using single-quoted strings with `\n`? Not high priority, though.

LarsKue reviewed (changes requested):
Very awesome changes, thank you for covering this so thoroughly!

I do see the need for some minor changes, particularly in the tests. See individual comments.

vpratz commented:
Thanks a lot for the review, I have addressed your comments. The tests are currently failing because I pulled in the changes from `dev`. If you have no objections, I will merge when the tests on `dev` are running again.

LarsKue commented:
@vpratz Yes, dev is now fixed. Can you please merge dev into this again and if tests pass, merge?
